app/src/main/java/ch/epfl/sdp/appart/database/local/LocalDatabase.java

```diff
@@ -0,0 +1,1055 @@
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while creating the ad folder !"));
+            return futureSuccess;
+        }
+
+
+        /*  I think we do not need to do something specific if the folder
+            already exists, we will just overwrite the files
+            Maybe we need to check for the number of images so that if the ad
+            has one image less than before we do not could keep a stale image
+            in the folder
+        */
+
+        //Building local versions of the ad only because we can build the
+        // card from the ad
+        Ad localAd = buildLocalAd(ad);
+        User localUser = buildLocalUser(user);
+        Card localCard = buildCardFromAd(localAd, cardId, adId,
+                localUser.getUserId());
+
+        //Adding the data to memory if we have done the first load
+        syncWithMemory(adId, localAd, localCard, localUser);
+
+        //Write the user
+        if (!writeUser(localUser)) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while writing the user !"));
+            return futureSuccess;
+        }
+
+        //We do not need to check if the set contains it or not.
+        this.userIds.add(localUser.getUserId());
+
+
+        //Serializing
+        if (!writeAd(adId, localAd, localUser.getUserId(), cardId)) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while writing the ad !"));
+            return futureSuccess;
+        }
+
+        if (!writeAdPhotos(adPhotos)) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while writing the ad photos !"));
+            return futureSuccess;
+        }
+
+        if (!writePanoramas(panoramas)) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while writing the ad photos !"));
+            return futureSuccess;
+        }
+        if (loadProfilePic != null && !localUser.hasDefaultProfileImage()) {
+            writeProfilePic(loadProfilePic, futureSuccess);
+        } else {
+            futureSuccess.complete(null);
+        }
+        return futureSuccess;
+    }
+
+    /**
+     * This method builds a card from the local ad, as most of the
+     * information a card contains is already in the ad.
+     *
+     * @param ad     the ad
+     * @param cardID the id of the card
+     * @param adID   the id of the ad
+     * @param userID the id of the user
+     * @return a card
+     */
+    private Card buildCardFromAd(Ad ad, String cardID, String adID,
+                                 String userID) {
+
+        String imageUrl = null;
+        if (!ad.getPhotosRefs().isEmpty()) {
+            imageUrl = ad.getPhotosRefs().get(0);
+        }
+
+        return new Card(cardID, adID, userID, ad.getCity(),
```

Comment:
Beep, Beep, I want more builders for these

Reply:
In another PR I think ;)